cannot compile with gcc-5.4

just pulled the master branch and I cannot compile it anymore with gcc-5.4. Is this a bug in my stl?
```
[ 67%] Building CXX object source/base/CMakeFiles/obj_base_release.dir/function_parser.cc.o
In file included from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_tempbuf.h:60:0,
                 from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_algo.h:62,
                 from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/algorithm:62,
                 from /scratch/opt/src/dealii/include/deal.II/base/table_indices.h:23,
                 from /scratch/opt/src/dealii/include/deal.II/base/tensor.h:21,
                 from /scratch/opt/src/dealii/include/deal.II/base/function.h:24,
                 from /scratch/opt/src/dealii/include/deal.II/base/auto_derivative_function.h:21,
                 from /scratch/opt/src/dealii/include/deal.II/base/function_parser.h:22,
                 from /scratch/opt/src/dealii/source/base/function_parser.cc:17:
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_construct.h: In instantiation of ‘void std::_Construct(_T1*, _Args&& ...) [with _T1 = std::unique_ptr<mu::Parser>; _Args = {const std::unique_ptr<mu::Parser, std::default_delete<mu::Parser> >&}]’:
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_uninitialized.h:75:18:   required from ‘static _ForwardIterator std::__uninitialized_copy<_TrivialValueTypes>::__uninit_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = __gnu_cxx::__normal_iterator<const std::unique_ptr<mu::Parser>*, std::vector<std::unique_ptr<mu::Parser> > >; _ForwardIterator = std::unique_ptr<mu::Parser>*; bool _TrivialValueTypes = false]’
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_uninitialized.h:126:15:   required from ‘_ForwardIterator std::uninitialized_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = __gnu_cxx::__normal_iterator<const std::unique_ptr<mu::Parser>*, std::vector<std::unique_ptr<mu::Parser> > >; _ForwardIterator = std::unique_ptr<mu::Parser>*]’
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_uninitialized.h:281:37:   required from ‘_ForwardIterator std::__uninitialized_copy_a(_InputIterator, _InputIterator, _ForwardIterator, std::allocator<_Tp>&) [with _InputIterator = __gnu_cxx::__normal_iterator<const std::unique_ptr<mu::Parser>*, std::vector<std::unique_ptr<mu::Parser> > >; _ForwardIterator = std::unique_ptr<mu::Parser>*; _Tp = std::unique_ptr<mu::Parser>]’
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_vector.h:322:31:   required from ‘std::vector<_Tp, _Alloc>::vector(const std::vector<_Tp, _Alloc>&) [with _Tp = std::unique_ptr<mu::Parser>; _Alloc = std::allocator<std::unique_ptr<mu::Parser> >]’
/usr/include/tbb/enumerable_thread_specific.h:477:49:   required from ‘tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::padded_element<U>::padded_element(const U&) [with U = std::vector<std::unique_ptr<mu::Parser> >; T = std::vector<std::unique_ptr<mu::Parser> >; Allocator = tbb::cache_aligned_allocator<std::vector<std::unique_ptr<mu::Parser> > >; tbb::ets_key_usage_type ETS_key_type = (tbb::ets_key_usage_type)1u]’
/usr/include/tbb/enumerable_thread_specific.h:636:91:   required from ‘T& tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::local(bool&) [with T = std::vector<std::unique_ptr<mu::Parser> >; Allocator = tbb::cache_aligned_allocator<std::vector<std::unique_ptr<mu::Parser> > >; tbb::ets_key_usage_type ETS_key_type = (tbb::ets_key_usage_type)1u; tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::reference = std::vector<std::unique_ptr<mu::Parser> >&]’
/usr/include/tbb/enumerable_thread_specific.h:602:25:   required from ‘T& tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::local() [with T = std::vector<std::unique_ptr<mu::Parser> >; Allocator = tbb::cache_aligned_allocator<std::vector<std::unique_ptr<mu::Parser> > >; tbb::ets_key_usage_type ETS_key_type = (tbb::ets_key_usage_type)1u; tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::reference = std::vector<std::unique_ptr<mu::Parser> >&]’
/scratch/opt/src/dealii/include/deal.II/base/thread_local_storage.h:210:23:   required from ‘T& dealii::Threads::ThreadLocalStorage<T>::get() [with T = std::vector<std::unique_ptr<mu::Parser> >]’
/scratch/opt/src/dealii/source/base/function_parser.cc:383:13:   required from ‘double dealii::FunctionParser<dim>::value(const dealii::Point<dim>&, unsigned int) const [with int dim = 1]’
/scratch/opt/src/dealii/source/base/function_parser.cc:478:16:   required from here
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_construct.h:75:7: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = mu::Parser; _Dp = std::default_delete<mu::Parser>]’
     { ::new(static_cast<void*>(__p)) _T1(std::forward<_Args>(__args)...); }
       ^
In file included from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/locale_conv.h:41:0,
                 from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/locale:43,
                 from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/iomanip:43,
                 from /scratch/opt/lib/trilinos/include/Epetra_ConfigDefs.h:152,
                 from /scratch/opt/lib/trilinos/include/Epetra_Comm.h:47,
                 from /scratch/opt/src/dealii/include/deal.II/base/utilities.h:28,
                 from /scratch/opt/src/dealii/include/deal.II/base/tensor.h:24,
                 from /scratch/opt/src/dealii/include/deal.II/base/function.h:24,
                 from /scratch/opt/src/dealii/include/deal.II/base/auto_derivative_function.h:21,
                 from /scratch/opt/src/dealii/include/deal.II/base/function_parser.h:22,
                 from /scratch/opt/src/dealii/source/base/function_parser.cc:17:
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/unique_ptr.h:356:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^
make[2]: *** [source/base/CMakeFiles/obj_base_release.dir/function_parser.cc.o] Error 1
make[1]: *** [source/base/CMakeFiles/obj_base_release.dir/all] Error 2
make: *** [all] Error 2
```

--- a/doc/news/changes/minor/20170726AlbertoSartori
+++ b/doc/news/changes/minor/20170726AlbertoSartori
@@ -1,0 +1,3 @@
+Fixed: FunctionParser uses shared_ptr instead of unique_ptr in order to fix compilation issues with gcc-5.4.
+<br>
+(Alberto Sartori, 2017/07/26)

--- a/doc/news/changes/minor/20170726AlbertoSartori
+++ b/doc/news/changes/minor/20170726AlbertoSartori
@@ -1,3 +1,5 @@
-Fixed: FunctionParser uses shared_ptr instead of unique_ptr in order to fix compilation issues with gcc-5.4.
+Fixed: The FunctionParser class demonstrated an incompatibility with
+(very) old versions of the Threading Building Blocks used in
+deal.II. This is now worked around.
 <br>
-(Alberto Sartori, 2017/07/26)
+(Alberto Sartori, Wolfgang Bangerth, 2017/07/26)

--- a/include/deal.II/base/function_parser.h
+++ b/include/deal.II/base/function_parser.h
@@ -328,11 +328,17 @@ private:
 
   /**
    * The muParser objects for each thread (and one for each component). We are
-   * storing shared_ptr so that we don't need to include the definition of
+   * storing a unique_ptr so that we don't need to include the definition of
    * mu::Parser in this header.
    */
-  // std::unique_ptr does not compile with gcc-5.4
+#if TBB_VERSION_MAJOR >= 4
+  mutable Threads::ThreadLocalStorage<std::vector<std::unique_ptr<mu::Parser> > > fp;
+#else
+  // older TBBs have a bug in which they want to return thread-local
+  // objects by value. this doesn't work for std::unique_ptr, so use a
+  // std::shared_ptr
   mutable Threads::ThreadLocalStorage<std::vector<std::shared_ptr<mu::Parser> > > fp;
+#endif
 
   /**
    * An array to keep track of all the constants, required to initialize fp in

--- a/include/deal.II/base/function_parser.h
+++ b/include/deal.II/base/function_parser.h
@@ -328,10 +328,11 @@ private:
 
   /**
    * The muParser objects for each thread (and one for each component). We are
-   * storing unique_ptr so that we don't need to include the definition of
+   * storing shared_ptr so that we don't need to include the definition of
    * mu::Parser in this header.
    */
-  mutable Threads::ThreadLocalStorage<std::vector<std::unique_ptr<mu::Parser> > > fp;
+  // std::unique_ptr does not compile with gcc-5.4
+  mutable Threads::ThreadLocalStorage<std::vector<std::shared_ptr<mu::Parser> > > fp;
 
   /**
    * An array to keep track of all the constants, required to initialize fp in

